app/src/main/java/com/nutomic/syncthingandroid/activities/FirstStartActivity.java

```diff
@@ -94,11 +123,28 @@ public boolean onTouch(View v, MotionEvent event) {
             }
         });
 
-        // Layouts of all welcome slides
-        mLayouts = new int[]{
-                R.layout.activity_firststart_slide1,
-                R.layout.activity_firststart_slide2,
-                R.layout.activity_firststart_slide3};
+        // Add welcome slides to be shown.
+        int[] colorsActive = getResources().getIntArray(R.array.array_dot_active);
+        int[] colorsInactive = getResources().getIntArray(R.array.array_dot_inactive);
+        int slideIndex = 0;
+        mSlides = new Slide[
+                1 +
+                (showSlideStoragePermission ? 1 : 0) +
+                (showSlideLocationPermission ? 1 : 0) +
+                (showSlideKeyGeneration ? 1 : 0)
+        ];
+        mSlides[slideIndex++] = new Slide(R.layout.activity_firststart_intro, colorsActive[0], colorsInactive[0]);
+        if (showSlideStoragePermission) {
+            mSlidePosStoragePermission = slideIndex;
+            mSlides[slideIndex++] = new Slide(R.layout.activity_firststart_storage_permission, colorsActive[1], colorsInactive[1]);
```

Comment:
You could add SlideType enum and then set the type for each slide, and find the index by iterating over slides later rather than storing some integer on the class body.

Reply:
You hear an improved implementation where slide type does not depend on the index of it, which makes the code easier to understand and reason wether it works correctly.

Reply:
Assigned for the next PR of @AudriusButkevicius . Merge and then do it, please. You've got my approval for this implementation.

Reply:
Sorry, but I am not going to blindly merge something regardless of how salty you behave.
I am trying to behave like a reasonable adult, but at some point I too will run out of patience.

Reply:
If you don't wish to drive this home, let's close this and part ways.

I'd prefer to have no contributions, over contributions where every improvement gets a toxic response for no apparent reason.

Reply:
The PR was tested on my phone, worked perfectly from the early beginning. Code shaping is yours, developing is mine.